[IRGen] Generate proper type descriptors for ObjC subscript accessors

jrose-apple opened the pull request:
[SR-8356](https://bugs.swift.org/browse/SR-8356) / rdar://problem/42284266

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -437,7 +437,8 @@ class ObjCProtocolInitializerVisitor
     if (prop->isSettable(nullptr)) {
-      emitObjCSetterDescriptorParts(IGM, prop, name, types, imp);
+      emitObjCSetterDescriptorParts(IGM, prop, /*extended*/true,
+                                    name, types, imp);
```
Protocols do, non-protocols don't. I don't know the story further than that, though; I just imitated what was already there. (That's why I tagged Greg for review, really.)

rjmccall replied:
Ah, I guess that makes as much sense as anything.  Could you make the fact that that's the reason behind the decision more explicit in the source?

jrose-apple replied:
Okay, I don't think this is correct after all. There's already a separate path for getting extended type encodings for protocols. @gparker42, how is this supposed to work?

jrose-apple committed:
[IRGen] Stop using "extended" type encodings where not supported

These are only used in two places in the Apple Objective-C runtime:

- A protocol's "extended method types" list
- Block type descriptors

We were using them when dynamically registering a protocol with the
Objective-C runtime, but that's just expecting "normal" types; the
"extended method types" list is never present in such a protocol.

jrose-apple committed:
[IRGen] Generate proper type descriptors for ObjC subscript accessors

https://bugs.swift.org/browse/SR-8356
rdar://problem/42284266

jrose-apple committed:
[IRGen] Stop registering protocol property accessors twice in JIT mode

We visited them twice, which led to registering them twice. Add a test
for this feature so that we don't regress on this or on the use of
non-extended types in the future (see previous commits).

jrose-apple commented:
\*sigh\* ObjCBool…

rjmccall reviewed (commented):
This seems reasonable to me.  What's going on with `ObjCBool`?

jrose-apple commented:
Oh, just that the test case I modified was already limited to `CPU=x86_64` because whoever wrote it originally didn't feel like conditionalizing it, but I forgot that that includes the 64-bit simulator. I think it's probably okay to just change it to OS=macosx, since it wasn't going to be testing most platforms anyway.

jrose-apple commented:
@swift-ci Please test
